source/custom/nixnet_service.custom.cpp

```diff
@@ -24,31 +24,42 @@ inline bool status_ok(int32 status)
 }
 
 // Helper to read fields of u32 returned by nxReadState of nxState_CANComm and set them on gRPC message
-void SetCanComm(const u32& input, nixnet_grpc::CanComm* output)
+void SetCanCommResponse(const u32& input, nixnet_grpc::CanCommResponse* output)
 {
-  output->set_comm_state(nxCANComm_Get_CommState(input));
+  u32 comm_state = nxCANComm_Get_CommState(input);
+  u32 last_error = nxCANComm_Get_LastErr(input);
+
+  output->set_comm_state(static_cast<nixnet_grpc::CanCommState>(comm_state));
+  output->set_comm_state_raw(comm_state);
   output->set_transceiver_error(nxCANComm_Get_TcvrErr(input));
   output->set_sleep(nxCANComm_Get_Sleep(input));
-  output->set_last_error(nxCANComm_Get_LastErr(input));
+  output->set_last_error(static_cast<nixnet_grpc::CanLastErr>(last_error));
+  output->set_last_error_raw(last_error);
   output->set_transmit_error_counter(nxCANComm_Get_TxErrCount(input));
   output->set_receive_error_counter(nxCANComm_Get_RxErrCount(input));
 }
 
 // Helper to read fields of u32 returned by nxReadState of nxState_FlexRayComm and set them on gRPC message
-void SetFlexRayComm(const u32& input, nixnet_grpc::FlexRayComm* output)
+void SetFlexRayCommResponse(const u32& input, nixnet_grpc::FlexRayCommResponse* output)
 {
-  output->set_poc_state(nxFlexRayComm_Get_POCState(input));
+  u32 poc_state = nxFlexRayComm_Get_POCState(input);
+
+  output->set_poc_state(static_cast<nixnet_grpc::FlexRayPocState>(poc_state));
+  output->set_poc_state_raw(poc_state);
   output->set_clock_correction_failed(nxFlexRayComm_Get_ClockCorrFailed(input));
   output->set_passive_to_active_count(nxFlexRayComm_Get_PassiveToActiveCount(input));
   output->set_channel_a_sleep(nxFlexRayComm_Get_ChannelASleep(input));
   output->set_channel_b_sleep(nxFlexRayComm_Get_ChannelBSleep(input));
 }
 
 // Helper to read two fields of u32 returned by nxReadState of nxState_LINComm and set them on gRPC message
-void SetLinComm(const u32* input, nixnet_grpc::LinComm* output)
+void SetLinCommResponse(const u32* input, nixnet_grpc::LinCommResponse* output)
```

Comment:
Sorry I missed this before. This is a strange way to pass a `u32` inputs. `const&` (used above) is more correct but also uncommon. Usually integers/numeric inputs are passed by value.

Make all of these the same. Preferably `u32 input` (by value) but `const u32& input` is OK.

Consider: you can also add a helper function to make these easier to unpack

```c++
template <typename T>
T get_raw_value(const void* pointer_to_raw_value)
{
   return *reinterpret_cast<const T*>(pointer_to_raw_value);
}

[...]
SetLinCommResponse(get_raw_value<u32>(state_value_raw), response->mutable_state_value()->mutable_lin_comm());

[...]

```

Reply:
I changed the two helpers above this to take in u32 inputs by value. But `SetLinCommResponse` wont change because the input is a u32 array and we need to pass a pointer. Regarding the `get_raw_value()` helper, we feel it would make the code less readable for the convert_to_grpc calls below, which would look something like : `get_raw_value<nxFlexRayStats>()` instead of just `(nxFlexRayStats_t*)`
Hence, I will skip the helper creation.